baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/DangerousParallelStreamUsageTest.java

```diff
@@ -0,0 +1,252 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.baseline.errorprone;
+
+import com.google.errorprone.CompilationTestHelper;
+import org.junit.Before;
+import org.junit.Test;
+
+public final class DangerousParallelStreamUsageTest {
+    private CompilationTestHelper compilationHelper;
+
+    @Before
+    public void before() {
+        compilationHelper = CompilationTestHelper.newInstance(DangerousParallelStreamUsage.class, getClass());
+    }
+
+    @Test
+    public void should_warn_when_parallel_with_no_arguments_is_invoked_on_subclass_of_java_stream() {
+        compilationHelper.addSourceLines(
+                "Test.java",
+                "import java.util.List;",
+                "import java.util.ArrayList;",
+                "class Test {",
+                "   public static final void main(String[] args) {",
+                "       List<String> list = new ArrayList<>();",
+                "       // BUG: Diagnostic contains: Should not use .parallel() on a Java stream.",
+                "       list.stream().parallel();",
+                "   }",
+                "}"
+        ).doTest();
+    }
+
+    @SuppressWarnings("MethodLength")
+    @Test
+    public void should_also_warn_when_parallel_with_arguments_is_invoked_on_subclass_of_java_stream() {
+        compilationHelper.addSourceLines(
+                "Test.java",
+                "import java.util.Comparator;",
+                "import java.util.Iterator;",
+                "import java.util.Optional;",
+                "import java.util.Spliterator;",
+                "import java.util.concurrent.ForkJoinPool;",
+                "import java.util.function.BiConsumer;",
+                "import java.util.function.BiFunction;",
+                "import java.util.function.BinaryOperator;",
+                "import java.util.function.Consumer;",
+                "import java.util.function.Function;",
+                "import java.util.function.IntFunction;",
+                "import java.util.function.Predicate;",
+                "import java.util.function.Supplier;",
+                "import java.util.function.ToDoubleFunction;",
+                "import java.util.function.ToIntFunction;",
+                "import java.util.function.ToLongFunction;",
+                "import java.util.stream.Collector;",
+                "import java.util.stream.DoubleStream;",
+                "import java.util.stream.IntStream;",
+                "import java.util.stream.LongStream;",
+                "import java.util.stream.Stream;",
+                "class Test {",
+                "   public static class FooStream<T> implements Stream<T> {",
```

Comment:
nit: There must be a simpler way to create a stream? Stream.of(1)?

Reply:
what I'm testing for here is subclassing a stream to add a method with name `parallel` but more arguments and make sure we fail for that too - the idea here is also banning `StreamEx.of(1).parallel(new ForkJoinPool())` for the reasons explained by James on our email thread (i.e. poor control over parallelism since the `ForkJoinPool` takes the number of cores in a system as an input to estimate how many tasks to run in parallel vs taking the max number of concurrent tasks explicitly)

Reply:
Cool, I'll defer to Dan if there's a better way to do this. Ideally your code would just say something like "when I call a method from this interface". This test is kinda asserting your understanding of this, but really it's testing if errorprone does the right thing.

Reply:
I agree its not ideal to have such a verbose test. I'm not sure of a much more succinct way of expressing this though

Reply:
I agree this test is questionable, however, I wanted to document the ban of `StreamEx#parallel(ForkJoinPool)` (or any variant of `parallel`) explicitly in a test - that method is not part of the `Stream` api - so in theory one could argue that they could have their own, correct, implementation (in reality they don't, they just defer to Java's parallel streams), so erring on the side of caution